internal/controllers/nmc_reconciler.go

```diff
@@ -555,104 +565,138 @@ func (h *nmcReconcilerHelperImpl) SyncStatus(ctx context.Context, nmcObj *kmmv1b
-	for nsn := range nodeModuleReadyLabels {
-		_, inSpec := specLabels[nsn]
-		_, inStatus := statusLabels[nsn]
-		if !inSpec && !inStatus {
-			meta.RemoveLabel(
-				&node,
-				utils.GetKernelModuleReadyNodeLabel(nsn.Namespace, nsn.Name),
-			)
+	// get spec labels and their config
+	specLabels := h.lph.getSpecLabelsAndTheirConfigs(nmc)
 
-			unloaded = append(unloaded, nsn)
-		}
-	}
+	// get status labels and their config
+	statusLabels := h.lph.getStatusLabelsAndTheirConfigs(nmc)
 
-	// v1 ready labels, deprecated - should be removed
-	for label := range deprecatedNodeModuleReadyLabels {
-		meta.RemoveLabel(&node, label)
-	}
+	// label in node but not in spec or status - should be removed
+	unloaded := h.removeOrphanedLabels(nodeModuleReadyLabels, specLabels, statusLabels)
 
 	// label in spec and status and config equal - should be added
-	for nsn, specConfig := range specLabels {
-		statusConfig, ok := statusLabels[nsn]
-		if ok && reflect.DeepEqual(specConfig, statusConfig) && !nodeModuleReadyLabels.Has(nsn) {
-			meta.SetLabel(
-				&node,
-				utils.GetKernelModuleReadyNodeLabel(nsn.Namespace, nsn.Name),
-				"",
-			)
+	loaded := h.addEqualLabels(nodeModuleReadyLabels, specLabels, statusLabels)
 
-			loaded = append(loaded, nsn)
-		}
-	}
+	labelsToBeLoaded := utils.ConvertNamespacedNameToStrings(loaded)
+	labelsToBeRemoved := utils.ConvertNamespacedNameToStrings(unloaded)
+	deprecatedLabels := utils.SetToSlice(deprecatedNodeModuleReadyLabels)
+
+	labelsToBeRemoved = append(labelsToBeRemoved, deprecatedLabels...)
 
-	if err := h.client.Patch(ctx, &node, patchFrom); err != nil {
-		return fmt.Errorf("could not patch node: %v", err)
+	if err := h.nodeAPI.UpdateLabels(ctx, node, labelsToBeLoaded, labelsToBeRemoved); err != nil {
+		return nil, nil, fmt.Errorf("could not patch node: %v", err)
 	}
 
-	for _, nsn := range unloaded {
+	return loaded, unloaded, nil
+}
+
+func (h *nmcReconcilerHelperImpl) RecordEvents(loadedModules []types.NamespacedName, unloadedModules []types.NamespacedName, node *v1.Node) {
+	for _, nsn := range loadedModules {
 		h.recorder.AnnotatedEventf(
-			&node,
+			node,
 			map[string]string{"module": nsn.String()},
 			v1.EventTypeNormal,
-			"ModuleUnloaded",
-			"Module %s unloaded from the kernel",
+			"ModuleLoaded",
+			"Module %s loaded into the kernel",
 			nsn.String(),
 		)
 	}
-
-	for _, nsn := range loaded {
+	for _, nsn := range unloadedModules {
 		h.recorder.AnnotatedEventf(
-			&node,
+			node,
 			map[string]string{"module": nsn.String()},
 			v1.EventTypeNormal,
-			"ModuleLoaded",
-			"Module %s loaded into the kernel",
+			"ModuleUnloaded",
+			"Module %s unloaded from the kernel",
 			nsn.String(),
 		)
 	}
+}
 
-	return nil
+func (h *nmcReconcilerHelperImpl) removeOrphanedLabels(nodeModuleReadyLabels sets.Set[types.NamespacedName],
+	specLabels map[types.NamespacedName]kmmv1beta1.ModuleConfig,
+	statusLabels map[types.NamespacedName]kmmv1beta1.ModuleConfig) []types.NamespacedName {
```

Comment:
```suggestion
	statusLabels map[types.NamespacedName]kmmv1beta1.ModuleConfig) []string {
```
Can we just return a label list instead? No reason to build an object just to parse it later on.

Reply:
following our meeting, I left it as it was